Image shared informers

soltysh opened the pull request:
This builds on top of the rebase PR, so only last commit matters here, but I wanted to start the discussion.

@smarterclayton mind taking a look if this is going in the right direction? While working on it I was wondering why we want to have two separate controllers one for scheduled and the other for regular image stream operations, isn't having a single sufficient?

soltysh commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(8) Use shared informer in image controllers](https://trello.com/c/7184zdCn/895-8-use-shared-informer-in-image-controllers)

soltysh commented:
@mfojtik fyi

mfojtik commented on pkg/cmd/server/origin/run_components.go:
```diff
@@ -428,19 +429,24 @@ func (c *MasterConfig) RunImageImportController() {
+		!c.Options.ImagePolicyConfig.DisableScheduledImport,
+	)
+	if c.Options.ImagePolicyConfig.DisableScheduledImport {
+		glog.V(2).Infof("Scheduled image import is disabled - the 'scheduled' flag on image streams will be ignored")
```
wonder if there is a way to communicate this to users (if I create scheduled import as a user, I have no idea it is disabled, right?)

soltysh replied:
The image stream controller could know that and could add an annotation with that info. I'll defer to @smarterclayton for opinion, since he has mixed feelings, but I agree it would be beneficial for users to know this.

soltysh commented:
[test]

@smarterclayton this is ready for final review

soltysh commented:
Flake https://github.com/openshift/origin/issues/14044

[test]

soltysh commented:
Flake https://github.com/openshift/origin/issues/14044

[test]

mfojtik commented on pkg/cmd/server/origin/run_components.go:
```diff
@@ -420,6 +420,7 @@ func (c *MasterConfig) RunServiceServingCertController(client kclientsetinternal
 
 // RunImageImportController starts the image import trigger controller process.
 func (c *MasterConfig) RunImageImportController() {
+	isInformer := c.Informers.ImageStreams()
```
nuke this,  you used this just once

mfojtik commented on pkg/cmd/server/origin/run_components.go:
```diff
@@ -431,19 +432,21 @@ func (c *MasterConfig) RunImageImportController() {
-	controller, scheduledController := factory.Create()
-	controller.Run()
+	ctrl, sched := imagecontroller.NewImageStreamControllers(
+		isInformer,
```
`c.Informers.ImageStreams()`

mfojtik commented on pkg/cmd/server/origin/run_components.go:
```diff
@@ -431,19 +432,21 @@ func (c *MasterConfig) RunImageImportController() {
-	}
-	controller, scheduledController := factory.Create()
-	controller.Run()
+	ctrl, sched := imagecontroller.NewImageStreamControllers(
```
s/ctrl/controller/

mfojtik commented on pkg/cmd/server/origin/run_components.go:
```diff
@@ -431,19 +432,21 @@ func (c *MasterConfig) RunImageImportController() {
 		glog.V(2).Infof("Scheduled image import is disabled - the 'scheduled' flag on image streams will be ignored")
 	} else {
-		scheduledController.RunUntil(utilwait.NeverStop)
+		sched.Run(utilwait.NeverStop)
```
scheduledController (hate cryptic short forms)

mfojtik commented on pkg/image/controller/factory.go:
```diff
@@ -3,144 +3,87 @@ package controller
-	q := cache.NewResyncableFIFO(cache.MetaNamespaceKeyFunc)
-	cache.NewReflector(lw, &api.ImageStream{}, q, f.ResyncInterval).Run()
+	// instantiate informer based image stream controller
+	ctrl := newImageStreamController(isInformer, isNamespacer)
```
controller

soltysh commented:
Flake https://github.com/openshift/origin/issues/13827

[test]

mfojtik commented:
@soltysh we might do this https://github.com/openshift/origin/pull/13996/commits/e486a4f13caf7263b3872aa7af78c04f41b8583d plumbing for the image controller as well (we are going to migrate all controllers eventually...) can you have a look? if not I can do that on top of this PR :-)

smarterclayton commented on pkg/image/controller/imagestream_controller.go:
```diff
@@ -25,25 +53,131 @@ type Notifier interface {
+
+	stream, err := c.getByKey(key.(string))
+	if err == nil && stream != nil {
+		glog.V(3).Infof("Queued import of stream %s/%s...", stream.Namespace, stream.Name)
```
Was this 3 before?

soltysh replied:
It's slightly different than what we've had before. But this is [this](https://github.com/openshift/origin/pull/13835/files#diff-0da76d25b3dc65e4fc80f1ab22aa745aL192) and it was 3 back then.

soltysh commented:
@mfojtik roger that, I'll try to update that tomorrow.

smarterclayton commented on pkg/image/controller/imagestream_controller.go:
```diff
@@ -25,25 +53,131 @@ type Notifier interface {
+	}
+	stream, err := c.lister.ImageStreams(namespace).Get(name, metav1.GetOptions{})
+	if apierrs.IsNotFound(err) {
+		glog.V(4).Infof("Image stream has been deleted: %v", key)
```
Doesn't need a log line

smarterclayton commented on pkg/image/controller/scheduled_image_controller.go:
```diff
@@ -0,0 +1,192 @@
+	if !needsScheduling(sharedStream) {
+		return ErrNotImportable
+	}
+	stream := *sharedStream
```
Not sufficient, this is mutating a shared cache

smarterclayton commented:
Logic looks ok except for a few things, want to see this rebased and green tests.

soltysh commented:
I've rebased and squashed all the changes to a single commit. @mfojtik where's that commit coming from, I can't find that code anywhere in origin :/ I've left there TODO. Let's merge this in on green tests and I'll work out a followup with the init changes, wdyt?

soltysh commented:
@mfojtik @smarterclayton any objections on merging this in?

mfojtik commented:
@soltysh i'm fine merging this but you should follow up with the controller initialization :-)

mfojtik commented:
[merge]

mfojtik commented:
[merge][severity: blocker] memory use is still p0 for 3.6

mfojtik commented:
[merge][severity: blocker]

flake: https://github.com/openshift/origin/issues/14349

smarterclayton commented:
Blocked by hate, force merging

mfojtik commented:
![evil-girl-fire-force-merge](https://cloud.githubusercontent.com/assets/44136/26465960/614bc0c6-418d-11e7-9d26-75a85007cd00.jpg)
